partition/timequery: Fix tiered storage timequery bug

wdberkeley opened the pull request:
This change refactors timequery logic to fix a rare bug where a timequery would return no result despite an offset matching the timestamp existing in cloud storage and local storage.

The bug is observed in the tiered storage model test when querying timestamps corresponding to offsets in the final, active segment of the log of a partition with tiered storage enabled and with aggressive cleanup settings. The partition gets into a state where the local log is truncated up to the high watermark, but the active segment remains. Querying for a timestamp within the bounds of the active segment meant that the timestamp was after the start timestamp for the local log, which comes from the base timestamp of the only segment, but the start offset of the log was after the max offset set by the timequery, which is the HWM and comes from the Kafka handler. This caused the timequery to return no result, incorrectly, without checking local or cloud storage.

The refactor fixes the bug and hopefully cleans up the logic a bit. Regression test included.

Fixes CORE-14569

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [X] v25.3.x
- [X] v25.2.x
- [X] v25.1.x
- [X] v24.3.x

## Release Notes

### Bug Fixes

* Fixed a rare bug that causes timequeries against partitions using tiered storage to incorrectly return no result when the partition's local log is empty but retains an active segment.

wdberkeley committed:
partition/timequery: Fix tiered storage timequery bug

This change refactors timequery logic to fix a rare bug where a
timequery would return no result despite an offset matching the
timestamp existing in cloud storage and local storage.

The bug is observed in the tiered storage model test when querying
timestamps corresponding to offsets in the final, active segment of the
log of a partition with tiered storage enabled and with aggressive
cleanup settings. The partition gets into a state where the local log is
truncated up to the high watermark, but the active segment remains.
Querying for a timestamp within the bounds of the active segment
meant that the timestamp was after the start timestamp for the local
log, which comes from the base timestamp of the only segment, but the
start offset of the log was after the max offset set by the timequery,
which is the HWM and comes from the Kafka handler. This caused the
timequery to return no result, incorrectly, without checking local or
cloud storage.

The refactor fixes the bug and hopefully cleans up the logic a bit.
Regression test included.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--0519d376-ecab-46c3-ade8-646f13ebf11b-->
<!--TEST RESULTS-->
<!--2068c2a0-cce8-4bdf-af8e-fffe190d13e5-->
<!--TEST RESULTS-->
<!--ace73614-c441-44c8-b414-17c9a7865475-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/76683">build#76683</a></summary>


| test_class                          | test_method                           | test_arguments              | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                         |
|:------------------------------------|:--------------------------------------|:----------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest               | test_higher_level_migration_api       | null                        | integration | https://buildkite.com/redpanda/redpanda/builds/76683#019a9e1b-6589-44f5-b422-f52f5bde80b2 | FLAKY         | 20/21    | upstream reliability is '99.41747572815534'. current run reliability is '95.23809523809523'. drift is 4.17938 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_higher_level_migration_api  |
| MountUnmountIcebergTest             | test_simple_remount                   | {"cloud_storage_type": 1}   | integration | https://buildkite.com/redpanda/redpanda/builds/76683#019a9e1d-ac19-465f-8f3e-32490d71688e | FLAKY         | 16/21    | upstream reliability is '79.19227392449517'. current run reliability is '76.19047619047619'. drift is 3.0018 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=MountUnmountIcebergTest&test_method=test_simple_remount            |
| MultiRestartTest                    | test_recovery_after_multiple_restarts | {"cloud_storage_type": 1}   | integration | https://buildkite.com/redpanda/redpanda/builds/76683#019a9e1d-ac24-4e3d-8e88-351b78aa9c2e | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=MultiRestartTest&test_method=test_recovery_after_multiple_restarts |
| WriteCachingFailureInjectionE2ETest | test_crash_all                        | {"use_transactions": false} | integration | https://buildkite.com/redpanda/redpanda/builds/76683#019a9e1d-ac20-4c51-be59-b163a0fd46ee | FLAKY         | 19/21    | upstream reliability is '91.27310061601642'. current run reliability is '90.47619047619048'. drift is 0.79691 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all     |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/76900">build#76900</a></summary>


| test_class       | test_method                           | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                         |
|:-----------------|:--------------------------------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| MultiRestartTest | test_recovery_after_multiple_restarts | {"cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/76900#019ab729-d35a-42d4-9277-b901ed71f32f | FLAKY         | 20/21    | upstream reliability is '95.89905362776025'. current run reliability is '95.23809523809523'. drift is 0.66096 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=MultiRestartTest&test_method=test_recovery_after_multiple_restarts |
| TimeQueryTest    | test_timequery_empty_local_log        | null                      | integration | https://buildkite.com/redpanda/redpanda/builds/76900#019ab729-3881-4fbd-b809-52cb49df1e98 | FLAKY         | 6/21     | upstream reliability is '100.0'. current run reliability is '28.57142857142857'. drift is 71.42857 and the allowed drift is set to 50. The test should FAIL            | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TimeQueryTest&test_method=test_timequery_empty_local_log           |
| TimeQueryTest    | test_timequery_empty_local_log        | null                      | integration | https://buildkite.com/redpanda/redpanda/builds/76900#019ab729-d359-48bd-a5b9-a8083ab8438d | FLAKY         | 6/21     | upstream reliability is '100.0'. current run reliability is '28.57142857142857'. drift is 71.42857 and the allowed drift is set to 50. The test should FAIL            | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TimeQueryTest&test_method=test_timequery_empty_local_log           |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77366">build#77366</a></summary>


| test_class                      | test_method                           | test_arguments                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                   | test_history                                                                                                                                                                 |
|:--------------------------------|:--------------------------------------|:--------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| JavaCompressionTest             | test_upgrade_java_compression         | {"compression_type": "gzip"}                      | integration | https://buildkite.com/redpanda/redpanda/builds/77366#019aeb99-9f85-49b7-85c5-d2f53bb579b5 | FLAKY         | 20/21    | upstream reliability is '88.20224719101124'. current run reliability is '95.23809523809523'. drift is -7.03585 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=JavaCompressionTest&test_method=test_upgrade_java_compression              |
| JavaCompressionTest             | test_upgrade_java_compression         | {"compression_type": "gzip"}                      | integration | https://buildkite.com/redpanda/redpanda/builds/77366#019aeb9a-2ee2-43b0-8f4f-d8c0077ad3b0 | FLAKY         | 20/21    | upstream reliability is '88.20224719101124'. current run reliability is '95.23809523809523'. drift is -7.03585 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=JavaCompressionTest&test_method=test_upgrade_java_compression              |
| JavaCompressionTest             | test_upgrade_java_compression         | {"compression_type": "lz4"}                       | integration | https://buildkite.com/redpanda/redpanda/builds/77366#019aeb99-9f87-4622-a59a-c6c09173769d | FLAKY         | 20/21    | upstream reliability is '88.23529411764706'. current run reliability is '95.23809523809523'. drift is -7.0028 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=JavaCompressionTest&test_method=test_upgrade_java_compression              |
| JavaCompressionTest             | test_upgrade_java_compression         | {"compression_type": "lz4"}                       | integration | https://buildkite.com/redpanda/redpanda/builds/77366#019aeb9a-2ee4-4b34-8ac3-7c952af48362 | FLAKY         | 20/21    | upstream reliability is '88.23529411764706'. current run reliability is '95.23809523809523'. drift is -7.0028 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=JavaCompressionTest&test_method=test_upgrade_java_compression              |
| JavaCompressionTest             | test_upgrade_java_compression         | {"compression_type": "snappy"}                    | integration | https://buildkite.com/redpanda/redpanda/builds/77366#019aeb99-9f88-4867-be14-f747b3ed8692 | FLAKY         | 20/21    | upstream reliability is '88.23529411764706'. current run reliability is '95.23809523809523'. drift is -7.0028 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=JavaCompressionTest&test_method=test_upgrade_java_compression              |
| JavaCompressionTest             | test_upgrade_java_compression         | {"compression_type": "snappy"}                    | integration | https://buildkite.com/redpanda/redpanda/builds/77366#019aeb9a-2ee5-4392-bfaf-261b3e833853 | FLAKY         | 19/21    | upstream reliability is '88.23529411764706'. current run reliability is '90.47619047619048'. drift is -2.2409 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=JavaCompressionTest&test_method=test_upgrade_java_compression              |
| JavaCompressionTest             | test_upgrade_java_compression         | {"compression_type": "zstd"}                      | integration | https://buildkite.com/redpanda/redpanda/builds/77366#019aeb99-9f8a-4e65-81b1-4fbb50a32e21 | FLAKY         | 20/21    | upstream reliability is '88.23529411764706'. current run reliability is '95.23809523809523'. drift is -7.0028 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=JavaCompressionTest&test_method=test_upgrade_java_compression              |
| JavaCompressionTest             | test_upgrade_java_compression         | {"compression_type": "zstd"}                      | integration | https://buildkite.com/redpanda/redpanda/builds/77366#019aeb9a-2ee7-4e3d-b834-478fdb10bc1b | FLAKY         | 20/21    | upstream reliability is '88.23529411764706'. current run reliability is '95.23809523809523'. drift is -7.0028 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=JavaCompressionTest&test_method=test_upgrade_java_compression              |
| NodesDecommissioningTest        | test_decommissioning_rebalancing_node | {"shutdown_decommissioned": false}                | integration | https://buildkite.com/redpanda/redpanda/builds/77366#019aeb9a-2ee7-4e3d-b834-478fdb10bc1b | FLAKY         | 15/21    | upstream reliability is '93.27731092436974'. current run reliability is '71.42857142857143'. drift is 21.84874 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_rebalancing_node |
| RedpandaNodeOperationsSmokeTest | test_node_ops_smoke_test              | {"cloud_storage_type": 1, "mixed_versions": true} | integration | https://buildkite.com/redpanda/redpanda/builds/77366#019aeb9a-2ee2-43b0-8f4f-d8c0077ad3b0 | FLAKY         | 13/21    | upstream reliability is '89.82300884955751'. current run reliability is '61.904761904761905'. drift is 27.91825 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test       |
</details>
<!--/body-->

nvartolomei commented on src/v/cluster/partition.cc:
```diff
@@ -609,58 +614,54 @@ partition::timequery(storage::timequery_config cfg) {
+      log()->from_log_offset(_raft->start_offset()),
+      local_query_cfg.min_offset);
+
+    co_return co_await local_timequery(local_query_cfg, false);
```
I have an intuition that this is unnecessary.

Equivalent and simpler _to comprehend_ (sample size = 1)?

```diff
 src/v/cluster/partition.cc | 15 ++-------------
 1 file changed, 2 insertions(+), 13 deletions(-)

diff --git i/src/v/cluster/partition.cc w/src/v/cluster/partition.cc
index 81296fe36c..d6b7b0d866 100644
--- i/src/v/cluster/partition.cc
+++ w/src/v/cluster/partition.cc
@@ -620,7 +620,7 @@ partition::timequery(storage::timequery_config cfg) {
     const bool local_covers_offsets = local_start_offset <= cfg.max_offset;
     const bool may_answer_from_local = local_covers_timestamp
                                        && local_covers_offsets;
-    if (may_answer_from_local) {
+    if (may_answer_from_local || !may_answer_from_cloud) {
         // The query is ahead of the local data's start_timestamp and
         // potentially overlaps with the local data offset range: this means it
         // _might_ hit on local data: start_timestamp is not precise, so once we
@@ -650,18 +650,7 @@ partition::timequery(storage::timequery_config cfg) {
     // 3. the local log start timestamp is after the timequery's timestamp.
     // If 1 or 2 hold, there is no offset to return. Otherwise, fall back
     // to a local timequery, which should return the start of the log.
-    if (may_answer_from_local || !local_covers_offsets) {
-        co_return std::nullopt;
-    }
-
-    // Adjust the lower bound for the local query as the min_offset
-    // corresponds to the full log (including tiered storage).
-    auto local_query_cfg = cfg;
-    local_query_cfg.min_offset = std::max(
-      log()->from_log_offset(_raft->start_offset()),
-      local_query_cfg.min_offset);
-
-    co_return co_await local_timequery(local_query_cfg, false);
+    co_return std::nullopt;
 }

 bool partition::may_read_from_cloud() const {
```

andrwng replied:
```
if (local_covers_offsets) {
  if (local_covers_timestamp) {
    auto res = co_await local_timequery(..., may_answer_from_cloud)
    if (res) {
      co_return res;
    }
    // Null is a signal that the timestamp is below the local log.
  }
  if (may_answer_from_cloud) {
    co_return co_await cloud_storage_timequery(...);
  }
  if (!local_covers_timestamp) {
    co_return nullopt;
  }
  co_return co_await local_timequery(..., false);
}
```

wdberkeley replied:
Yeah, I think this makes sense. Thanks.

nvartolomei reviewed (dismissed):
Approving. As far as the fix goes it seems correct. The comment is a suggestion.

andrwng commented on src/v/cluster/partition.cc:
```diff
@@ -609,58 +614,54 @@ partition::timequery(storage::timequery_config cfg) {
-            // The local storage hit a case where it needs to fall back
-            // to querying cloud storage.
-            co_return co_await cloud_storage_timequery(cfg);
         }
```
nit: can you add a comment about the intentional fallthrough? And that a nullopt from the local_timequery is a signal to check cloud

wdberkeley replied:
I think applying NV's suggestion makes this all clearer.

andrwng commented on src/v/cluster/partition.cc:
```diff
@@ -609,58 +614,54 @@ partition::timequery(storage::timequery_config cfg) {
+        // _might_ hit on local data: start_timestamp is not precise, so once we
+        // query we might still fall back to cloud storage
         //
         // We also need to adjust the lower bound for the local query as the
         // min_offset corresponds to the full log (including tiered storage).
         auto local_query_cfg = cfg;
         local_query_cfg.min_offset = std::max(
-          log()->from_log_offset(_raft->start_offset()),
-          local_query_cfg.min_offset);
-
-        // If the min_offset is ahead of max_offset, the local log is empty
-        // or was truncated since the timequery_config was created.
-        if (local_query_cfg.min_offset > local_query_cfg.max_offset) {
-            co_return std::nullopt;
-        }
+          local_start_offset, local_query_cfg.min_offset);
```
In the previous code, it was clearer that we would only ever call local_timequery() or cloud_storage_timequery() once. Is it a bug that we aren't preserving that?

andrwng replied:
Actually I guess that isn't true since there are cases where we called local_timequery() and then cloud_storage_timequery().

I'm finding the new code structure a bit confusing since it at first seems like we can call local_timequery() twice. But because the condition here is `may_answer_from_local` and we are conditioning on that below to return a nullopt, that isn't possible.

wdberkeley replied:
With NV's suggestion we maintain this property.

andrwng commented on src/v/cluster/partition.cc:
```diff
@@ -609,58 +614,54 @@ partition::timequery(storage::timequery_config cfg) {
-        // If the min_offset is ahead of max_offset, the local log is empty
-        // or was truncated since the timequery_config was created.
-        if (local_query_cfg.min_offset > local_query_cfg.max_offset) {
-            co_return std::nullopt;
```
Just making sure I'm seeing the bug through the refactor, would an equivalent fix have been to replace this line with this?

```
if (may_answer_from_cloud) {
  co_return co_await cloud_storage_timequery(cfg);
}
co_return std::nullopt;
```

wdberkeley replied:
See NV's suggestion.

wdberkeley committed:
partition/timequery: Simplify tq logic further

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/76900-->
<!--977292c1-adfc-4ea5-93d6-0854776c95ff-->
<!--c30b5eaf-d3b1-470e-bd61-0fbffd2e899e-->
<!--title-->
## Retry command for [Build#76900](https://buildkite.com/redpanda/redpanda/builds/76900)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
skip-redpanda-build
skip-units
skip-rebase
tests/rptest/tests/timequery_test.py::TimeQueryTest.test_timequery_empty_local_log
```
<!--/body-->

nvartolomei commented:
```
DEBUG 2025-11-24 19:18:20,762 [shard 1:kafk] cluster - partition.cc:683 - timequery (raft) {kafka/tqtopic/0} cfg(k)={min_offset: 3072, max_offset: 3071, time:{timestamp: 1764098234000}, type_filter:batch_type::raft_data}
WARN  2025-11-24 19:18:20,763 [shard 0:kafk] kafka - connection_context.cc:1117 - Error processing request: std::runtime_error (ntp {kafka/tqtopic/0}: data offset 3071 is outside the translation range (starting at 3072))

```

nvartolomei commented:
I can swear this test passed locally when I proposed the [patch](https://github.com/redpanda-data/redpanda/pull/28642#discussion_r2547681428).

Anyhow:

https://github.com/redpanda-data/redpanda/blob/69e52106e68a7ebe58b4e5311a7d97d73a2c626a/src/v/storage/offset_translator_state.cc#L42-L47

We need a special case for empty log. I guess we found out what the original code was attempting to do ... but incorrectly. We need the special case.

wdberkeley committed:
cluster/partition: Add timequery empty range check to handle empty log

It can happen that a timequery hits an empty local log and its max
offset is before the min offset, which is clamped to the start of the
log. This will fail translation. This change adds extra early returns to
local and cloud timequeries to handle this special case pre-translation.

wdberkeley commented:
Force push to rebase on dev, then a push to add a special case for the empty log, where max_offset can end up before min_offset (which is clamped to the start_offset).

vbotbuildovich commented:
/backport v25.3.x

vbotbuildovich commented:
/backport v25.2.x

vbotbuildovich commented:
/backport v25.1.x

vbotbuildovich commented:
/backport v24.3.x
